File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Python/javonet/core/protocol/TypeDeserializer.py
import struct

from javonet.utils import Command, CommandType, RuntimeName
from javonet.utils.StringEncodingMode import StringEncodingMode


class TypeDeserializer:

    @staticmethod
    def deserialize_command(encoded_command ):
        return Command(RuntimeName(encoded_command[0]), CommandType(encoded_command[1]), [])

    @staticmethod
    def deserialize_string(string_encoding_mode, encoded_string):
        if string_encoding_mode == StringEncodingMode.ASCII:
            return bytearray(encoded_string).decode('ascii')
        if string_encoding_mode == StringEncodingMode.UTF8:
            return bytearray(encoded_string).decode('utf-8')
        if string_encoding_mode == StringEncodingMode.UTF16:
            return bytearray(encoded_string).decode('utf-16')
        if string_encoding_mode == StringEncodingMode.UTF32:
            return bytearray(encoded_string).decode('utf-32')

        raise IndexError("String encoding mode out of range")

    @staticmethod
    def deserialize_int(encoded_int):
        return struct.unpack("<i", bytearray(encoded_int))[0]
    
    @staticmethod
    def deserialize_bool(encoded_bool):
        return struct.unpack("<?", bytearray(encoded_bool))[0]
    
    @staticmethod
    def deserialize_float(encoded_float):
        return struct.unpack("<f", bytearray(encoded_float))[0]

    @staticmethod
    def deserialize_byte(encoded_byte):
        return struct.unpack("<B", bytearray(encoded_byte))[0]
    
    @staticmethod
    def deserialize_char(encoded_char):
        return struct.unpack("<b", bytearray(encoded_char))[0]

    @staticmethod
    def deserialize_longlong(encoded_longlong):
        return struct.unpack("<q", bytearray(encoded_longlong))[0]

    @staticmethod
    def deserialize_double(encoded_double):
        return struct.unpack("<d", bytearray(encoded_double))[0]

    @staticmethod
    def deserialize_ullong(encoded_unsigned_longlong):
        return struct.unpack("<Q", bytearray(encoded_unsigned_longlong))[0]

    @staticmethod
    def deserialize_uint(encoded_unsigned_int):
        return struct.unpack("<I", bytearray(encoded_unsigned_int))[0]

    @staticmethod
    def deserialize_none(encoded_none):
        return None

File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Nodejs/lib/core/protocol/TypeDeserializer.js
const Command = require('../../utils/Command')
const StringEncodingMode = require('../../utils/StringEncodingMode')
const { Buffer } = require('buffer')

class TypeDeserializer {
    static deserializeCommand(encodedCommand) {
        return new Command(encodedCommand[0], encodedCommand[1], [])
    }

    static deserializeString(stringEncodingMode, encodedString) {
        switch (stringEncodingMode) {
            case StringEncodingMode.ASCII:
                return String.fromCharCode(...encodedString)
            case StringEncodingMode.UTF8: {
                const decoder = new TextDecoder('UTF-8')
                return decoder.decode(new Uint8Array(encodedString))
            }
            case StringEncodingMode.UTF16: {
                let str = ''
                let newBuffer = new Uint8Array(encodedString)
                for (let i = 0; i < newBuffer.length; i++) {
                    newBuffer[i] = encodedString[i]
                }
                for (let i = 0; i < encodedString.length; i = i + 2) {
                    str += String.fromCharCode(newBuffer[i] + 256 * newBuffer[i + 1])
                }
                return str
            }
            case StringEncodingMode.UTF32:
                throw 'Type utf32-encoded string not supported in JavaScript'
            default:
                throw 'Unknown string encoding - not supported in JavaScript'
        }
    }

    static deserializeInt(encodedInt) {
        return (
            (encodedInt[0] & 0xff) |
            ((encodedInt[1] & 0xff) << 8) |
            ((encodedInt[2] & 0xff) << 16) |
            ((encodedInt[3] & 0xff) << 24)
        )
    }

    static deserializeBool(encodedBool) {
        return encodedBool[0] === 1
    }

    static deserializeFloat(encodedFloat) {
        return Buffer.from(encodedFloat).readFloatLE()
    }

    static deserializeByte(encodedByte) {
        return Buffer.from(encodedByte).readUint8()
    }

    static deserializeChar(encodedChar) {
        return Buffer.from(encodedChar).readUint8()
    }

    static deserializeLongLong(encodedLongLong) {
        return Buffer.from(encodedLongLong).readBigInt64LE()
    }

    static deserializeDouble(encodedDouble) {
        return Buffer.from(encodedDouble).readDoubleLE()
    }

    static deserializeULLong(encodedULLong) {
        return Buffer.from(encodedULLong).readBigUInt64LE()
    }

    static deserializeUInt(encodedUInt) {
        return Buffer.from(encodedUInt).readUIntLE(0, 4)
    }

    // eslint-disable-next-line no-unused-vars
    static deserializeNull(encodedNull = null) {
        return null
    }
}

module.exports = TypeDeserializer
